### ISSUE TITLE
ValidationException: Missing required files to delete

### ISSUE THREAD

**Issue by bk-mz** [apache/iceberg#9741]
### Apache Iceberg version

1.4.0

### Query engine

Spark

### Please describe the bug 🐞

Hey folks, I'm running a streaming app with 'MERGE INTO' sink. Same time I'm scheduling rewrite_data_files and expire_snapshots procedures in the background.

Code:

```java
public class IcebergTableStreamer extends BaseTableConfig implements TableWriteStreamer {
    private final SparkSession sparkSession;

    public IcebergTableStreamer(TableDefinition tableDefinition,
                                RuntimeConfig runtimeConfig,
                                SparkSession sparkSession) {
        super(tableDefinition, runtimeConfig);
        this.sparkSession = sparkSession;

        var scheduler = Executors.newSingleThreadScheduledExecutor();
        scheduler.scheduleAtFixedRate(this::compact, 0, 10, TimeUnit.MINUTES);
        scheduler.scheduleAtFixedRate(this::expireSnapshots, 0, 1, TimeUnit.DAYS);

        // todo rewrite manifests
    }

    @Override
    public void stream(DataStreamWriter<Row> streamWriter) throws TimeoutException, StreamingQueryException {
        streamWriter
            .option("checkpointLocation", checkpointPath)
            .option("mergeSchema", "true")
            .option("fanout-enabled", "true")
            .outputMode(OutputMode.Append())
            .foreachBatch(this::batch)
            .start()
            .awaitTermination();
    }

    private void batch(Dataset<Row> batch, Long batchId) {
        // deduplicate by precombine field
        var window = Window
            .partitionBy(col(tableDefinition.key()))
            .orderBy(col(tableDefinition.precombineField()).desc());
        batch
            .withColumn("row_num", row_number().over(window))
            .filter("row_num = 1")
            .drop("row_num")
            .createOrReplaceTempView("batch");

        // ugly hack to create table if it does not exist.
        var tmpSpark = batch.sparkSession();
        String createTable = String.format("""
            CREATE TABLE if not exists %s.%s
            USING iceberg
            PARTITIONED BY (year, month, day, hour)
            LOCATION '%s'
            TBLPROPERTIES ('write.spark.accept-any-schema'='true')
            AS SELECT * FROM batch
            """, databaseName, tableName, tablePath);
        tmpSpark.sql(createTable);

        var sql = String.format("""
                merge into %s.%s base_table
                using batch
                on
                    base_table.%s       = batch.%s
                when MATCHED and (batch.%s > base_table.%s) then UPDATE SET *
                when not MATCHED then INSERT *
                """,
            databaseName, tableName,
            tableDefinition.key(), tableDefinition.key(),
            tableDefinition.precombineField(), tableDefinition.precombineField()
        );
        tmpSpark.sql(sql);
    }

    private void compact() {
        try {
            // todo probably does not make sense to rewrite_data_files for current hour.
            sparkSession.sql(String.format("CALL glue.system.rewrite_data_files(table => '%s.%s')",
                databaseName, tableName)).show(false);
        } catch (Exception e) {
            System.out.println(e);
        }
    }

    private void expireSnapshots() {
        try {
            var table = databaseName + "." + tableName;
            var olderThan = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss.SSS")
                .format(LocalDateTime.now().minusDays(1));
            sparkSession.sql("CALL glue.system.expire_snapshots(table => '" + table +
                    "', older_than => TIMESTAMP '" + olderThan + "')")
                .show(false);
        } catch (Exception e) {
            System.out.println(e);
        }
    }

}
```

Job is running for 4-5 batches, then fails with exception below:

```Caused by: org.apache.iceberg.exceptions.ValidationException: Missing required files to delete: <files.parquet>

	at org.apache.iceberg.exceptions.ValidationException.check(ValidationException.java:49) ~[iceberg-spark3-runtime.jar:?]

	at org.apache.iceberg.ManifestFilterManager.validateRequiredDeletes(ManifestFilterManager.java:237) ~[iceberg-spark3-runtime.jar:?]

	at org.apache.iceberg.ManifestFilterManager.filterManifests(ManifestFilterManager.java:199) ~[iceberg-spark3-runtime.jar:?]

	at org.apache.iceberg.MergingSnapshotProducer.apply(MergingSnapshotProducer.java:824) ~[iceberg-spark3-runtime.jar:?]

	at org.apache.iceberg.BaseOverwriteFiles.apply(BaseOverwriteFiles.java:31) ~[iceberg-spark3-runtime.jar:?]

	at org.apache.iceberg.SnapshotProducer.apply(SnapshotProducer.java:226) ~[iceberg-spark3-runtime.jar:?]

	at org.apache.iceberg.BaseOverwriteFiles.apply(BaseOverwriteFiles.java:31) ~[iceberg-spark3-runtime.jar:?]

	at org.apache.iceberg.SnapshotProducer.lambda$commit$2(SnapshotProducer.java:376) ~[iceberg-spark3-runtime.jar:?]

	at org.apache.iceberg.util.Tasks$Builder.runTaskWithRetry(Tasks.java:413) ~[iceberg-spark3-runtime.jar:?]

	at org.apache.iceberg.util.Tasks$Builder.runSingleThreaded(Tasks.java:219) ~[iceberg-spark3-runtime.jar:?]

	at org.apache.iceberg.util.Tasks$Builder.run(Tasks.java:203) ~[iceberg-spark3-runtime.jar:?]

	at org.apache.iceberg.util.Tasks$Builder.run(Tasks.java:196) ~[iceberg-spark3-runtime.jar:?]

	at org.apache.iceberg.SnapshotProducer.commit(SnapshotProducer.java:374) ~[iceberg-spark3-runtime.jar:?]

	at org.apache.iceberg.BaseOverwriteFiles.commit(BaseOverwriteFiles.java:31) ~[iceberg-spark3-runtime.jar:?]

	at org.apache.iceberg.spark.source.SparkWrite.commitOperation(SparkWrite.java:222) ~[iceberg-spark3-runtime.jar:?]

	at org.apache.iceberg.spark.source.SparkWrite.access$1300(SparkWrite.java:84) ~[iceberg-spark3-runtime.jar:?]

	at org.apache.iceberg.spark.source.SparkWrite$CopyOnWriteOperation.commitWithSerializableIsolation(SparkWrite.java:446) ~[iceberg-spark3-runtime.jar:?]

	at org.apache.iceberg.spark.source.SparkWrite$CopyOnWriteOperation.commit(SparkWrite.java:416) ~[iceberg-spark3-runtime.jar:?]

	at org.apache.spark.sql.execution.datasources.v2.V2TableWriteExec.writeWithV2(WriteToDataSourceV2Exec.scala:422) ~[spark-sql_2.12-3.4.1-amzn-2.jar:?]

	at org.apache.spark.sql.execution.datasources.v2.V2TableWriteExec.writeWithV2$(WriteToDataSourceV2Exec.scala:382) ~[spark-sql_2.12-3.4.1-amzn-2.jar:?]

	at org.apache.spark.sql.execution.datasources.v2.ReplaceDataExec.writeWithV2(ReplaceDataExec.scala:29) ~[iceberg-spark3-runtime.jar:?]

	at org.apache.spark.sql.execution.datasources.v2.V2ExistingTableWriteExec.run(WriteToDataSourceV2Exec.scala:360) ~[spark-sql_2.12-3.4.1-amzn-2.jar:?]

	at org.apache.spark.sql.execution.datasources.v2.V2ExistingTableWriteExec.run$(WriteToDataSourceV2Exec.scala:359) ~[spark-sql_2.12-3.4.1-amzn-2.jar:?]

	at org.apache.spark.sql.execution.datasources.v2.ReplaceDataExec.run(ReplaceDataExec.scala:29) ~[iceberg-spark3-runtime.jar:?]

	at org.apache.spark.sql.execution.datasources.v2.V2CommandExec.result$lzycompute(V2CommandExec.scala:43) ~[spark-sql_2.12-3.4.1-amzn-2.jar:?]

	at org.apache.spark.sql.execution.datasources.v2.V2CommandExec.result(V2CommandExec.scala:43) ~[spark-sql_2.12-3.4.1-amzn-2.jar:?]

	at org.apache.spark.sql.execution.datasources.v2.V2CommandExec.executeCollect(V2CommandExec.scala:49) ~[spark-sql_2.12-3.4.1-amzn-2.jar:?]

	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.$anonfun$applyOrElse$1(QueryExecution.scala:104) ~[spark-sql_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.catalyst.QueryPlanningTracker$.withTracker(QueryPlanningTracker.scala:107) ~[spark-catalyst_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.execution.SQLExecution$.withTracker(SQLExecution.scala:250) ~[spark-sql_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.execution.SQLExecution$.executeQuery$1(SQLExecution.scala:123) ~[spark-sql_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$9(SQLExecution.scala:160) ~[spark-sql_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.catalyst.QueryPlanningTracker$.withTracker(QueryPlanningTracker.scala:107) ~[spark-catalyst_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.execution.SQLExecution$.withTracker(SQLExecution.scala:250) ~[spark-sql_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$8(SQLExecution.scala:160) ~[spark-sql_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:271) ~[spark-sql_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:159) ~[spark-sql_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:827) ~[spark-sql_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:69) ~[spark-sql_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.applyOrElse(QueryExecution.scala:101) ~[spark-sql_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.applyOrElse(QueryExecution.scala:97) ~[spark-sql_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformDownWithPruning$1(TreeNode.scala:554) ~[spark-catalyst_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:107) ~[spark-catalyst_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDownWithPruning(TreeNode.scala:554) ~[spark-catalyst_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.org$apache$spark$sql$catalyst$plans$logical$AnalysisHelper$$super$transformDownWithPruning(LogicalPlan.scala:32) ~[spark-catalyst_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning(AnalysisHelper.scala:267) ~[spark-catalyst_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning$(AnalysisHelper.scala:263) ~[spark-catalyst_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:32) ~[spark-catalyst_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:32) ~[spark-catalyst_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDown(TreeNode.scala:530) ~[spark-catalyst_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.execution.QueryExecution.eagerlyExecuteCommands(QueryExecution.scala:97) ~[spark-sql_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.execution.QueryExecution.commandExecuted$lzycompute(QueryExecution.scala:84) ~[spark-sql_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.execution.QueryExecution.commandExecuted(QueryExecution.scala:82) ~[spark-sql_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.Dataset.<init>(Dataset.scala:221) ~[spark-sql_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.Dataset$.$anonfun$ofRows$2(Dataset.scala:101) ~[spark-sql_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:827) ~[spark-sql_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.Dataset$.ofRows(Dataset.scala:98) ~[spark-sql_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.SparkSession.$anonfun$sql$1(SparkSession.scala:640) ~[spark-sql_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:827) ~[spark-sql_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:630) ~[spark-sql_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:671) ~[spark-sql_2.12-3.4.1-amzn-2.jar:3.4.1-amzn-2]

	at mycoolstreamingapp.IcebergTableStreamer.batch(IcebergTableStreamer.java:93) ~[__app__.jar:0.0.1-SNAPSHOT]```

Please help me understand if this is expected behavior and thereby what's wrong with my table services that I schedule in parallel OR if this is something indeed of a bug.

Catalog Settings:
```java
                "spark.sql.defaultCatalog", "glue",
                "spark.sql.catalog.glue", "org.apache.iceberg.spark.SparkCatalog",
                "spark.sql.catalog.glue.catalog-impl", "org.apache.iceberg.aws.glue.GlueCatalog",
                "spark.sql.catalog.glue.io-impl", "org.apache.iceberg.aws.s3.S3FileIO",
                "spark.sql.extensions", "org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions"       
```

**Comment by fengguangyuan** [comment]
Hi, there.
I believe it's the protection for the correctness of the existed data, instead of a bug.

> Basic logics of parallel write: possibly read the same data, but never commit metadata based on the same snapshot.

From line `at org.apache.iceberg.BaseOverwriteFiles.apply(BaseOverwriteFiles.java:31) ~[iceberg-spark3-runtime.jar:?]
` in the stacktrace, we can know the thread is overwriting files but failed with losing delete files.

Considering Compact & Overwrite tasks running in parallel (Expire task does nothing), it's possible that they are holding the same snapshot (including the same view of the delete files) to do their works, but at some point a Compact task `committed before the Overwrite task `trying to call the internal method to commit metadata, obviously these overwrite tasks will fail with `ValidationException` on the latest snapshot (old delete files are invisible) during in committing.

Hope I had explained some key points to help you to understand the commit logic in parallel. :)

**Comment by bk-mz** [comment]
@fengguangyuan 

>but at some point a Compact task committed before the Overwrite task trying to call the internal method to commit metadata, obviously these overwrite tasks will fail with ValidationException

How come optimistic concurrency controls are not dealing with this? Does Iceberg have to, is it turned on automatically or do I have to set up things myself?

**Comment by fengguangyuan** [comment]
> @fengguangyuan
> 
> > but at some point a Compact task committed before the Overwrite task trying to call the internal method to commit metadata, obviously these overwrite tasks will fail with ValidationException
> 
> How come optimistic concurrency controls are not dealing with this? Does Iceberg have to, is it turned on automatically or do I have to set up things myself?

Question 1: OCC is only for `CommitFailedException` with 4 retries, but fail immediately for other cases. 
Question 2: Maybe `base_table` is in `COPY_ON_WRITE` mode? Please try to change the table property to `MERGE_ON_READ`.
```java
  public static final String MERGE_MODE = "write.merge.mode";
  // public enum RowLevelOperationMode {
  //    COPY_ON_WRITE("copy-on-write"),
  //    MERGE_ON_READ("merge-on-read");
  // }
  public static final String MERGE_MODE_DEFAULT = RowLevelOperationMode.COPY_ON_WRITE.modeName();
```
